### PR TITLE
Dynamic target pause

### LINKED ISSUE #1408: Support pause properly in multi-user environment (for operator)

### PR DESCRIPTION
**Pull request by Razz4780**
Resolves #1408 

Clients can now dynamically request the agent to pause the remote container. This is done with a new message, which is sent only once (by the CLI, right after fetching the remote env). The remote container remains paused by the agent until exit or an unpause request has been received (this case will be handled in the operator).

The flag controlling the pause feature is moved from `AgentConfig` to `LayerConfig` and is still accessible through `MIRRORD_PAUSE` env variable.

### CODE REVIEW

**Comment by Razz4780** [comment]
> > this PR is not `draft` only to enable CI
> 
> You can allow the repo's workflows to run in your fork, then you can see the CI runs in your [fork's Actions tab](https://github.com/Razz4780/mirrord/actions). We can also just enable the workflow for drafts, bring it to discussion with the team if you think it should be enabled.

Oh, that's cool. Didn't know about it, thanks!

File: mirrord/agent/src/main.rs [hunk 1]
```diff
@@ -71,93 +74,78 @@ mod watched_task;
...
         let container =
             get_container(args.container_id.as_ref(), args.container_runtime.as_ref()).await?;
 
-        if container.is_none() && args.pause {
-            Err(AgentError::MissingContainerInfo)?
-        }
+        let handle = match container {
+            Some(container) => Some(ContainerHandle::new(container).await?),
+            None => None,
+        };
```
[the remark this answers is not in the document]
Comment (Razz4780) [reply mirrord/agent/src/main.rs:97]: I'd leave this as it is, working with both futures and `Result`s/`Option`s is always pretty verbose

File: mirrord/cli/src/execution.rs [hunk 2]
```diff
@@ -189,6 +178,32 @@ impl MirrordExecution {
         })
     }
 
+    /// Retrieve remote environment from the connected agent.
+    async fn get_remote_env(
```
[the remark this answers is not in the document]
Comment (Razz4780) [reply mirrord/cli/src/execution.rs:182]: Did you mean `tracing::instrument`? It seems there's no use of that across the `cli` crate

[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord/cli/src/execution.rs:182]: Yeah.

Oh snap, should we start using it there?

Anyway, suggestion remains for things outside `cli` crate.

**Comment by Razz4780** [comment]
> Some minor stuff.
> 
> As I understand this feature is still ongoing some discussion if it should be agent or operator?

The final decision is to move the refcount logic to the operator, converted to draft

File: mirrord/agent/src/main.rs [hunk 3]
```diff
@@ -71,93 +74,78 @@ mod watched_task;
...
-                "Agent exiting without having removed all the clients. Unpausing target container."
-            );
-            if let Err(err) = executor::block_on(self.container.as_ref().unwrap().unpause()) {
-                error!(
-                    "Could not unpause target container while exiting early, got error: {err:?}"
-                );
-            }
-        }
-    }
+    /// Handle to the target container if there is one.
```
Comment (meowjesty) [inline mirrord/agent/src/main.rs:81]: I would keep the docs justifying why this is an `Option`.

File: mirrord/agent/src/main.rs [hunk 4]
```diff
@@ -71,93 +74,78 @@ mod watched_task;
...
-    /// the container info (runtime and id) was not.
+    /// Return [`Err`] if container runtime operations failed.
     pub async fn new(args: &Args) -> Result<State> {
         let container =
             get_container(args.container_id.as_ref(), args.container_runtime.as_ref()).await?;
 
-        if container.is_none() && args.pause {
-            Err(AgentError::MissingContainerInfo)?
-        }
+        let handle = match container {
```
Comment (meowjesty) [inline mirrord/agent/src/main.rs:94]: As far as I can see, we don't use the original `container` after this, so why not:

```suggestion
        let container = match container {
```

**Review by meowjesty — CHANGES_REQUESTED** [review]
From what I understand, the part of multi client pause is now handled in the operator (or will be, not sure if it's supported there yet)?

:play_or_pause_button:

File: mirrord/agent/src/container_handle.rs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+#[derive(Debug)]
+struct Inner {
+    /// The managed container.
+    container: Container,
+    /// Cached process ID of the container.
+    pid: u64,
+    /// Cached environment of the container.
+    raw_env: HashMap<String, String>,
+    /// Whether the container is paused.
+    paused: RwLock<bool>,
```
[the remark this answers is not in the document]
Comment (Razz4780) [reply mirrord/agent/src/container_handle.rs:21]: Using `RwLock` here was just much more simple than figuring out how to handle weird scenarios only with an atomic bool. Like: one user requests pause and other user requests unpause at the same time, two users request pause but the pause request fails, etc.

**Review by meowjesty — APPROVED** [review]
:bear: :+1:

**Pull request merged**